### PR TITLE
Fix issues with activity table

### LINKED ISSUE #56: On the view activity cell, user pictures are wacky
Some things that happen:
- Activity cell has two different pictures
- Tap on activity and view it - there is only one user!
- Go back to the table view and look at the cell again. it now has two pictures, but they're both the same.

There should be only one. I haven't been able to reliably reproduce this, but it happens very frequently. I think you could reproduce it by tapping into and out of a lot of activities.

### LINKED ISSUE #87: Flicker on Discover / My Activity
This is due to reloading the table view in both ViewDidLoad() and ViewWillAppear(), so yeah. We (me Jake and Grant) talked about just having time stamp and refreshing after set amount of time (like 10 minutes???)

### LINKED ISSUE #97: Data isn't consistent throughout apps, causing lost updates
If you have multiple devices, often one of them will have stale data and seems disconnected from Firebase.  It will not update when people join or leave activities.  

This is annoying for views not updating if your device is "disconnected", but also causes a big problem for other users.

1. Join an activity with a "connected" device on an account
2. Join the same activity with a "disconnected" device on a different account
3. The first account's membership in the activity is overwritten by the second account.

### LINKED ISSUE #111: Josh is the only user
IMG_0429

### LINKED ISSUE #113: Activity View Flashes Continuously
It seems that when visiting the activity view, after a while it just keeps reloading the whole thing. This means that all the users just flash and flash and flash. Additionally, when it flashes, it attempts to scroll to the top of the table view, which also does weird stuff to the FAB when it happens.

issue

### LINKED ISSUE #117: Activities are blank
There's an issue where, when navigating to an activity, it will just show a blank screen.

### LINKED ISSUE #118: Location & Date info text are switched up in edit activity

### PR DESCRIPTION
**Pull request by noahtallen**
This PR:

- Fixes #113
- Fixes #87
- Fixes #56
- Fixes #97
- Fixes #111
- Fixes #117
- Fixes #118
- [EDIT: not #119, but it rarely happens because other bugs are fixed]
- Adds "pull to refresh" to the activity table
- Shows activity indicator while results are loading

### To Test:
- Run the app and examine all of the activity tables
- Make sure that updates and deletions to activities properly refresh on the tableview in realtime.
- Make sure that there is no inconsistency with user pictures - the ones on the table should be the same as in view Activity.
- Make sure that the user experience isn't jarring - we don't want to have a lot of flashes and stuff.

### CODE REVIEW

File: Buddies/Activities/ActivityList.swift [hunk 1]
```diff
@@ -0,0 +1,126 @@
...
+
+    func hasNoActivities() -> Bool {
+        return activities.flatMap { $0 }.count == 0
+    }
+
+    func get(_ indexPath: IndexPath) -> Activity {
+        return activities[indexPath.section][indexPath.row]
+    }
+
+    func setActivities(_ activities: [[Activity]]) -> [IndexPath]? {
```
Comment (noahtallen) [inline Buddies/Activities/ActivityList.swift:47]: Hi I'm Luke.  It's a little confusing that `setActivities` returns indices for removed activities, but nothing else.  Maybe just a different name?

File: Buddies/Activities/ActivityList.swift [hunk 2]
```diff
@@ -0,0 +1,126 @@
...
+    // If we don't handle section headers in this class, they won't display whatsoever:
+    func tableView(_ tableView: UITableView, titleForHeaderInSection section: Int) -> String? {
+        if section < sectionHeaders.count {
+            return sectionHeaders[section]
+        }
+        return nil
+    }
+
+    // Returns the index path of the matching ID:
+    func getIndexPath(of activityId: ActivityId, in section: Int) -> IndexPath? {
```
Comment (noahtallen) [inline Buddies/Activities/ActivityList.swift:94]: HOT

File: Buddies/Activities/ActivityList.swift [hunk 3]
```diff
@@ -0,0 +1,126 @@
...
+            return IndexPath(row: index, section: section)
+        } else {
+            return nil
+        }
+    }
+
+    // Gets the index path of an activityID no matter its section:
+    func getIndexPath(of activityId: ActivityId) -> IndexPath? {
+        for (i, section) in self.activities.enumerated() {
+            for (j, activity) in section.enumerated() {
```
Comment (noahtallen) [inline Buddies/Activities/ActivityList.swift]: You can just call `getIndexPath(of activityId: ActivityId, in section: Int)` inside of your outer for loop, I think.

File: Buddies/Activities/ActivityList.swift [hunk 4]
```diff
@@ -0,0 +1,126 @@
...
+                    return IndexPath(row: j, section: i)
+                }
+            }
+        }
+        return nil
+    }
+
+    // Updates the given activity and returns its index path:
+    func updateAndGetIndexPath(of activity: Activity, at section: Int) -> IndexPath? {
+        if (section >= activities.count) {
```
Comment (noahtallen) [inline Buddies/Activities/ActivityList.swift]: Could use `getIndexPath(of activityId: ActivityId, in section: Int)` to get indexPath, then update based on it, then return it.

File: Buddies/Activities/View Activity/ActivityDescriptionController.swift [hunk 5]
```diff
@@ -79,6 +79,13 @@ class ActivityDescriptionController: UIView, UICollectionViewDataSource, UIColle
     
     let topicDataSource = TopicStubDataSource()
 
+    override func layoutSubviews() {
+        super.layoutSubviews()
+        miniUser1.makeCircle()
```
Comment (noahtallen) [inline Buddies/Activities/View Activity/ActivityDescriptionController.swift:84]: MAYBE GET RID OF THIS IDK IM LUJKE

File: Buddies/Models/Activity.swift [hunk 6]
```diff
@@ -130,4 +131,17 @@ class Activity {
             self.members.append(uid)
         }
     }
+    
+    func isLooselyDifferent(from otherActivity: Activity?) -> Bool {
+        if let otherActivity = otherActivity {
```
Comment (noahtallen) [inline Buddies/Models/Activity.swift]: This isn't used

File: Buddies/Models/User.swift [hunk 7]
```diff
@@ -68,6 +68,16 @@ class OtherUser : User {
                          bio: bio,
                          favoriteTopics: favoriteTopics)
     }
+    
+    func isLooselyDifferent(from otherUser: User?) -> Bool {
```
Comment (noahtallen) [inline Buddies/Models/User.swift]: Isn't used

File: Buddies/Models/User.swift [hunk 8]
```diff
@@ -211,4 +221,14 @@ class LoggedInUser : User {
                     notificationToken: data["notification_token"] as? String,
                     chatReadAt: chatReadAt)
     }
+    
+    func isLooselyDifferent(from otherUser: User?) -> Bool {
+        if let otherUser = otherUser {
```
Comment (noahtallen) [inline Buddies/Models/User.swift]: You know ;)

File: Buddies/Utilities/DataAccessor.swift [hunk 9]
```diff
@@ -142,35 +138,61 @@ class DataAccessor : LoggedInUserInvalidationDelegate, ActivityInvalidationDeleg
...
+        var users = [User]()
+        var didFinishSetup = false
         let cancelers = userIds.map { uid in useUser(id: uid) { user in
-            if let user = user {
-                users[user.uid] = user
+            if let user = user, handledUsers[uid] == nil {
+                // User is not loaded yet, and the user exists
+                users.append(user)
+                handledUsers[uid] = true
+            } else if handledUsers[uid] == nil {
```
Comment (noahtallen) [inline Buddies/Utilities/DataAccessor.swift:173]: This isn't right, says Noah

~ Luke

File: Buddies/Utilities/DataAccessor.swift [hunk 10]
```diff
@@ -331,11 +325,12 @@ class DataAccessor : LoggedInUserInvalidationDelegate, ActivityInvalidationDeleg
     func onInvalidateActivity(activity: Activity?, id: String) {
         if let updatedActivity = activity {
             _activityCache.setObject(updatedActivity, forKey: updatedActivity.activityId as AnyObject)
+            _activityListeners[id]?.forEach { $0.fn(updatedActivity) }
         } else {
             // Activity was deleted, or is invalid:
             _activityCache.removeObject(forKey: id as AnyObject)
+            _activityListeners[id]?.forEach { $0.fn(activity) }
         }
-        _activityListeners[id]?.forEach { $0.fn(activity) }
```
Comment (noahtallen) [inline Buddies/Utilities/DataAccessor.swift]: We can just do this line, instead of twice above.

File: Buddies/Utilities/DataAccessor.swift (hunk 9 above)
Comment (noahtallen) [reply Buddies/Utilities/DataAccessor.swift:173]: lol this is actually fine

**Review by lukemeierj — APPROVED** [review]

**Pull request merged**